StabilityTracer: MobileNotes at WebKit: WebKit::AuxiliaryProcessProxy::protectedConnection const

RupinMittal opened the pull request:
#### 8d4bcdd23b51a62872c9c72896235b33b20dafd1
<pre>
StabilityTracer: MobileNotes at WebKit: WebKit::AuxiliaryProcessProxy::protectedConnection const
<a href="https://rdar.apple.com/145887512">rdar://145887512</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=290763">https://bugs.webkit.org/show_bug.cgi?id=290763</a>

Reviewed by NOBODY (OOPS!).

WebPageProxy::willPerformPasteCommand() queues an async IPC to
the Network Process. It&apos;s ok if the Network Process hasn&apos;t launched
yet because the IPC is queued.

But using waitForAsyncReplyAndDispatchImmediately requires the
Network Process to be fully launched. We&apos;re hitting the release
assert in AuxiliaryProcessProxy::protectedConnection because
the Network Process hasn&apos;t fully launched at the time of use.

But we don&apos;t need to hang the UI Process here until it recievies
a response from the Network Process because there are no synchronous
operations happening right after it. So we can simply remove the
call to waitForAsyncReplyAndDispatchImmediately.

So now WebPageProxy sends an IPC to the Network Process to grant
the Web Process file access for that file URL and the Network Process
will send IPC to the Web Process once access is granted.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::executeEditCommand):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d4bcdd23b51a62872c9c72896235b33b20dafd1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97585 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17210 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7426 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102689 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48114 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17504 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25663 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74354 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31536 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100588 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88244 "Found 10 new API test failures: TestWebKitAPI.PasteImage.PastePNGFile, TestWebKitAPI.PasteMixedContent.ImageFileAndPlainText, TestWebKitAPI.PasteMixedContent.ImageFileAndHTML, TestWebKitAPI.PasteMixedContent.ImageFileAndURL, TestWebKitAPI.PasteMixedContent.ImageFileAndWebArchive, TestWebKitAPI.PasteMixedContent.ImageFileWithHTMLAndURL, TestWebKitAPI.PasteImage.PasteTIFFFile, TestWebKitAPI.PasteMixedContent.ImageFileAndRTF, TestWebKitAPI.PasteImage.PasteJPEGFile, TestWebKitAPI.PasteImage.PasteGIFFile (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54700 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47556 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6208 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104692 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24665 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17994 "Found 1 new test failure: imported/w3c/web-platform-tests/workers/data-url-shared.html (failure)") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25037 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84374 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82826 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27375 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5062 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18259 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24626 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29795 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24448 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27762 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26022 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0c9844351d919441d2561ae2b97eb116a73d41ef. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43319

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3550,8 +3550,11 @@ void WebPageProxy::executeEditCommand(const String& commandName, const String& a
-            protectedWebsiteDataStore()->protectedNetworkProcess()->protectedConnection()->waitForAsyncReplyAndDispatchImmediately<Messages::NetworkProcess::AllowFilesAccessFromWebProcess>(*replyID, 100_ms);
+        if (auto replyID = willPerformPasteCommand(*pasteAccessCategory, WTFMove(completionHandler), frameID)) {
+            Ref networkProcess = protectedWebsiteDataStore()->networkProcess();
+            if (networkProcess->hasConnection())
```
I'm not super familiar with Pascoe's original PR but I suspect we want to somehow queue an IPC to the NetworkProcess to let it know the WebProcess is allowed as soon as it finishes launching.

Note that `AuxiliaryProcessProxy::send()`  for example has logic to queue messages if the underlying process is still launching. This is why we almost never send IPC directly using an IPC::Connection object when dealing with an AuxiliaryProcessProxy.

cdumez replied:
The UIProcess needs to send the `WebPage::ExecuteEditCommand` IPC to the WebContent process but can only do so after we've sent the `NetworkProcess::AllowFilesAccessFromWebProcess` IPC to the network process and we've received confirmation it has received that IPC.

This is because once we send the `WebPage::ExecuteEditCommand` IPC to the WebProcess, it may send IPC to the network process to access local files (and it should be allowed).

As long as we still enforce this ordering, we should be fine. The trick is that we cannot send the `NetworkProcess::AllowFilesAccessFromWebProcess` IPC to the network process until after it is finished launching (and thus has an IPC::Connection). But as long as we delay the `WebPage::ExecuteEditCommand` IPC sending to the WebProcess as well, we should be OK.

charliewolfe replied:
This isn’t where the IPC to the NP is being sent; this is just hanging the UIP before calling the completion handler.

pascoej replied:
The issue this code solves is that there are tests (and API clients) that run `[webView paste:nil]` and then expect to be able to immediately run something like `EXPECT_WK_STREQ("image/tiff", [webView stringByEvaluatingJavaScript:@"file = dataTransfer.files[0]; file.type"]);` and have it work. 

Before `AllowFilesAccessFromWebProcess ` was introduced this worked because the paste command would queue a message to the WebProcess that would be processed by the time the next `stringByEvaluatingJavaScript` gets processed by the web process. However, when introducing `AllowFilesAccessFromWebProcess,` without this `waitForAsyncReplyAndDispatchImmediately` the web process would not have gotten the original paste message by the time the `stringByEvaluatingJavaScript` gets processed (because the UI process sends the `stringByEvaluatingJavaScript` before the `AllowFilesAccessFromWebProcess` message to the network process gets back.)


I think bar-another way to maintain this ordering, we need a way to wait until the NetworkProcess has fully launched to send the message. I think this might be `networkProcessLaunchingIfNecessary,` but since Chris said that queues messages up it still might not suffice. It might make sense to queue the `AllowFilesAccessFromWebProcess` but, but immediately call the completion handler in `executeEditCommand` if the network process is still launching since any messages from the WebProcess to the network process would be queued as well.

achristensen07 replied:
If there is no network process connection, can't we just remember the pid and paths in the UI process then send them to the network process when we do initialize it, kind of like we do with NetworkProcessCreationParameters.allowedFirstPartiesForCookies?

RupinMittal committed:
StabilityTracer: MobileNotes at WebKit: WebKit::AuxiliaryProcessProxy::protectedConnection const
rdar://145887512
https://bugs.webkit.org/show_bug.cgi?id=290763

Reviewed by NOBODY (OOPS!).

WebPageProxy::willPerformPasteCommand() queues an async IPC to
the Network Process. It's ok if the Network Process hasn't launched
yet because the IPC is queued.

But using waitForAsyncReplyAndDispatchImmediately requires the
Network Process to be fully launched. We're hitting the release
assert in AuxiliaryProcessProxy::protectedConnection because
the Network Process hasn't fully launched at the time of use.

But we don't need to hang the UI Process here until it recievies
a response from the Network Process because there are no synchronous
operations happening right after it. So we can simply remove the
call to waitForAsyncReplyAndDispatchImmediately.

So now WebPageProxy sends an IPC to the Network Process to grant
the Web Process file access for that file URL and the Network Process
will send IPC to the Web Process once access is granted.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::executeEditCommand):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b7a2543ecb4e781f941beb5fa540820eeda59ad5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43319

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8d4bcdd23b51a62872c9c72896235b33b20dafd1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43319

RupinMittal commented:
This was fixed in https://commits.webkit.org/305173@main.
